File: src/net/dinkla/raytracer/math/Vector2D.java
package net.dinkla.raytracer.math;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:17:35
 */
public class Vector2D extends Element2D {

    public Vector2D(final float x, final float y) {
        super(x, y);
    }

    public Vector2D(final double x, final double y) {
        super(x, y);
    }

    public Vector2D plus(final Vector2D v) {
        return new Vector2D(x + v.x, y + v.y);
    }

    public Vector2D minus(final Vector2D v) {
        return new Vector2D(x - v.x, y - v.y);
    }

    public Vector2D mult(final float s) {
        return new Vector2D(s * x, s * y);
    }

    public float dot(final Vector2D v)  {
        return x * v.x + y * v.y;
    }

    public float dot(final Normal v)  {
        return x * v.x + y * v.y;
    }

    public Vector2D normalize() {
        final float l = length();
        return new Vector2D(x / l, y / l);
    }

}

File: src/commonMain/kotlin/net/dinkla/raytracer/math/Vector2D.kt
package net.dinkla.raytracer.math

import kotlin.math.sqrt

data class Vector2D(
    val x: Double,
    val y: Double,
) {
    operator fun plus(v: Vector2D) = Vector2D(x + v.x, y + v.y)

    operator fun minus(v: Vector2D) = Vector2D(x - v.x, y - v.y)

    operator fun times(s: Double) = Vector2D(s * x, s * y)

    infix fun dot(v: Vector2D): Double = x * v.x + y * v.y

    infix fun dot(v: Normal): Double = x * v.x + y * v.y

    operator fun unaryMinus() = Vector2D(-x, -y)

    val length: Double
        get() = sqrt(x * x + y * y)

    fun normalize(): Vector2D {
        val len = length
        return Vector2D(x / len, y / len)
    }

    override fun toString(): String = "($x,$y)"
}

operator fun Double.times(v: Vector2D) = Vector2D(this * v.x, this * v.y)
